packages/mdc-toolbar/foundation.js

```diff
@@ -0,0 +1,234 @@
+    this.keyHeights.flexibleExpansionRatio = flexibleRowMaxRatio - 1;
+    this.keyHeights.maxTranslateYRatio = this.fixedLastrow_? this.keyHeights.toolbarRatio - flexibleRowMaxRatio : 0;
+    this.keyHeights.scrollThesholdRatio = (this.fixedLastrow_? this.keyHeights.toolbarRatio : flexibleRowMaxRatio) - 1;
+  }
+
+  getRowHeight_() {
+    const breakpoint = MDCToolbarFoundation.numbers.TOOLBAR_MOBILE_BREAKPOINT;
+    return this.adapter_.getViewportWidth() <= breakpoint ?
+      MDCToolbarFoundation.numbers.TOOLBAR_ROW_MOBILE_HEIGHT : MDCToolbarFoundation.numbers.TOOLBAR_ROW_HEIGHT;
+  }
+
+  scrolledOutOfTheshold_(scrollTop) {
+    if (scrollTop > this.keyHeights.scrollTheshold) {
+      if (this.executedLastChange_) {
+        return true;
+      }
+      this.executedLastChange_ = true;
+    } else {
+      this.executedLastChange_ = false;
+    }
+    return false;
+  }
+
+  setKeyHeights_({isInit = false} = {}) {
+    const newToolbarRowHeight = this.getRowHeight_();
+    if (newToolbarRowHeight !== this.keyHeights.toolbarRowHeight || isInit) {
+      this.keyHeights.toolbarRowHeight = newToolbarRowHeight;
+      this.keyHeights.toolbarHeight = this.keyHeights.toolbarRatio * this.keyHeights.toolbarRowHeight;
+      this.keyHeights.flexibleExpansionHeight =
+        this.keyHeights.flexibleExpansionRatio * this.keyHeights.toolbarRowHeight;
+      this.keyHeights.maxTranslateYDistance = this.keyHeights.maxTranslateYRatio * this.keyHeights.toolbarRowHeight;
+      this.keyHeights.scrollTheshold = this.keyHeights.scrollThesholdRatio * this.keyHeights.toolbarRowHeight;
+      this.changeAdjustElementStyles();
+      this.changeToolbarStyles();
+    }
+  }
+
+  changeToolbarFlexibleState_(flexibleExpansionRatio) {
+    this.adapter_.removeClass(MDCToolbarFoundation.cssClasses.FLEXIBLE_MAX);
+    this.adapter_.removeClass(MDCToolbarFoundation.cssClasses.FLEXIBLE_MIN);
+    if (flexibleExpansionRatio === 1) {
+      this.adapter_.addClass(MDCToolbarFoundation.cssClasses.FLEXIBLE_MAX);
+    } else if (flexibleExpansionRatio === 0) {
+      this.adapter_.addClass(MDCToolbarFoundation.cssClasses.FLEXIBLE_MIN);
+    }
+  }
+
+  changeToolbarFixedState_(scrollTop) {
+    const translateDistance = Math.max(0, Math.min(
+      scrollTop - this.keyHeights.flexibleExpansionHeight,
+      this.keyHeights.maxTranslateYDistance));
+    this.adapter_.setStyleForRootElement('transform', `translateY(${-translateDistance}px)`);
+    if (translateDistance === this.keyHeights.maxTranslateYDistance) {
+      this.adapter_.addClass(MDCToolbarFoundation.cssClasses.FIXED_AT_LAST_ROW);
+    } else {
+      this.adapter_.removeClass(MDCToolbarFoundation.cssClasses.FIXED_AT_LAST_ROW);
+    }
+  }
+
+  changeFlexibleRowElementStyles_(flexibleExpansionRatio) {
+    if (this.fixedLastrow_ || this.fixedAllRow_) {
+      const height = this.keyHeights.flexibleExpansionHeight * flexibleExpansionRatio;
+      this.adapter_.setStyleForFlexibleRowElement('height',
```

Comment:
Ah, not _crazy_ about using height here. But given that everything needs to seamlessly move with the flexible header, I don't see a way around it. Plus, it doesn't seem to affect performance _that_ much, save for chrome on android. But even then, it still does work. It'll be interesting to see how this holds up on production sites with more complex UI than what's found in our demos.